src/ejson2/EjsonEncoder.ts

```diff
@@ -0,0 +1,593 @@
+    writer.u8(0x7b); // {
+    this.writeStr('$binary');
+    writer.u8(0x3a); // :
+    writer.u8(0x7b); // {
+    this.writeStr('base64');
+    writer.u8(0x3a); // :
+    this.writeStr(this.uint8ArrayToBase64(value.data));
+    writer.u8(0x2c); // ,
+    this.writeStr('subType');
+    writer.u8(0x3a); // :
+    this.writeStr(value.subtype.toString(16).padStart(2, '0'));
+    writer.u8(0x7d); // }
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonCodeAsEjson(value: BsonJavascriptCode): void {
+    // Write {"$code":"..."}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$code');
+    writer.u8(0x3a); // :
+    this.writeStr(value.code);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonCodeWScopeAsEjson(value: BsonJavascriptCodeWithScope): void {
+    // Write {"$code":"...","$scope":{...}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$code');
+    writer.u8(0x3a); // :
+    this.writeStr(value.code);
+    writer.u8(0x2c); // ,
+    this.writeStr('$scope');
+    writer.u8(0x3a); // :
+    this.writeAny(value.scope);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonSymbolAsEjson(value: BsonSymbol): void {
+    // Write {"$symbol":"..."}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$symbol');
+    writer.u8(0x3a); // :
+    this.writeStr(value.symbol);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonTimestampAsEjson(value: BsonTimestamp): void {
+    // Write {"$timestamp":{"t":...,"i":...}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$timestamp');
+    writer.u8(0x3a); // :
+    writer.u8(0x7b); // {
+    this.writeStr('t');
+    writer.u8(0x3a); // :
+    this.writeNumber(value.timestamp);
+    writer.u8(0x2c); // ,
+    this.writeStr('i');
+    writer.u8(0x3a); // :
+    this.writeNumber(value.increment);
+    writer.u8(0x7d); // }
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonDbPointerAsEjson(value: BsonDbPointer): void {
+    // Write {"$dbPointer":{"$ref":"...","$id":{...}}}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$dbPointer');
+    writer.u8(0x3a); // :
+    writer.u8(0x7b); // {
+    this.writeStr('$ref');
+    writer.u8(0x3a); // :
+    this.writeStr(value.name);
+    writer.u8(0x2c); // ,
+    this.writeStr('$id');
+    writer.u8(0x3a); // :
+    this.writeAny(value.id);
+    writer.u8(0x7d); // }
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonMinKeyAsEjson(): void {
+    // Write {"$minKey":1}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$minKey');
+    writer.u8(0x3a); // :
+    this.writeNumber(1);
+    writer.u8(0x7d); // }
+  }
+
+  private writeBsonMaxKeyAsEjson(): void {
+    // Write {"$maxKey":1}
+    const writer = this.writer;
+    writer.u8(0x7b); // {
+    this.writeStr('$maxKey');
+    writer.u8(0x3a); // :
+    this.writeNumber(1);
+    writer.u8(0x7d); // }
```

Comment:
In this method (and throughout this file) do not use `.writeStr()` for constant values. If value is known ahead of time use a combination of `writer.u32()`, `writer.u16()`, and `writer.u8()` to write binary octets directly.